src/analysis/analytic/composition.lean

```diff
@@ -0,0 +1,647 @@
+the source of the change of variables (`comp_partial_source`), its target
+(`comp_partial_target`) and the change of variables itself (`comp_change_of_variables`) before
+giving the main statement in `comp_partial_sum`. -/
+
+/-- Source set in the change of variables to compute the composition of partial sums of formal
+power series -/
+def comp_partial_sum_source (N : ℕ) : finset (Σ n, (fin n) → ℕ) :=
+finset.sigma (finset.range N) (λ (n : ℕ), fintype.pi_finset (λ (i : fin n), finset.Ico 1 N) : _)
+
+@[simp] lemma mem_comp_partial_sum_source_iff (N : ℕ) (i : Σ n, (fin n) → ℕ) :
+  i ∈ comp_partial_sum_source N ↔ i.1 < N ∧ ∀ (a : fin i.1), 1 ≤ i.2 a ∧ i.2 a < N :=
+by simp [comp_partial_sum_source]
+
+/-- Change of variables appearing to compute the composition of partial sums of formal
+power series -/
+def comp_change_of_variables (N : ℕ) (i : Σ n, (fin n) → ℕ) (hi : i ∈ comp_partial_sum_source N) :
+  (Σn, composition n) :=
+begin
+  rcases i with ⟨n, f⟩,
+  rw mem_comp_partial_sum_source_iff at hi,
+  exact ⟨finset.univ.sum f, list.of_fn (λ a, ⟨f a, (hi.2 a).1⟩), by simp [list.of_fn_sum]⟩
+end
+
+@[simp] lemma comp_change_of_variables_length
+  (N : ℕ) {i : Σ n, (fin n) → ℕ} (hi : i ∈ comp_partial_sum_source N) :
+  composition.length (comp_change_of_variables N i hi).2 = i.1 :=
+begin
+  rcases i with ⟨k, blocks_fun⟩,
+  dsimp [comp_change_of_variables],
+  simp only [composition.length, composition.blocks, list.map_of_fn, list.length_of_fn]
+end
+
+lemma comp_change_of_variables_blocks_fun
+  (N : ℕ) {i : Σ n, (fin n) → ℕ} (hi : i ∈ comp_partial_sum_source N) (j : fin i.1) :
+  (comp_change_of_variables N i hi).2.blocks_fun
+    ⟨j.val, (comp_change_of_variables_length N hi).symm ▸ j.2⟩ = i.2 j :=
+begin
+  rcases i with ⟨n, f⟩,
+  dsimp [composition.blocks_fun, composition.blocks, comp_change_of_variables],
+  simp only [list.map_of_fn, pnat.mk_coe, list.nth_le_of_fn', function.comp_app],
+  apply congr_arg,
+  rw fin.ext_iff
+end
+
+/-- Target set in the change of variables to compute the composition of partial sums of formal
+power series, here given a a set. -/
+def comp_partial_sum_target_set (N : ℕ) : set (Σ n, composition n) :=
+{i | (i.2.length < N) ∧ (∀ (j : fin i.2.length), i.2.blocks_fun j < N)}
+
+lemma comp_partial_sum_target_subset_image_comp_partial_sum_source
+  (N : ℕ) (i : Σ n, composition n) (hi : i ∈ comp_partial_sum_target_set N) :
+  ∃ j (hj : j ∈ comp_partial_sum_source N), i = comp_change_of_variables N j hj :=
+begin
+  rcases i with ⟨n, c⟩,
+  simp [comp_partial_sum_target_set] at hi,
+  refine ⟨⟨c.length, c.blocks_fun⟩, _, _⟩,
+  { simp only [mem_comp_partial_sum_source_iff, hi.left, hi.right, true_and, and_true],
+    exact λ a, c.one_le_blocks' _ },
+  { dsimp [comp_change_of_variables],
+    rw composition.sigma_eq_iff_blocks_eq,
+    simp [composition.blocks_fun, composition.blocks, ← list.map_of_fn, - list.map_of_fn, composition.length],
```

Comment:
This caught my eye because of the nonterminal `simp`, but even more catching is the pattern ` ← list.map_of_fn, - list.map_of_fn`. I haven't used backward simp much; is this saying that if you add the backward direction of something that's in the default simp set, it doesn't remove the forward direction automatically? That sounds like a clear mistake on the part of `simp`.

Reply:
I'm pretty sure this was explicitly in the design. I know this used to be done with `foo.symm, -foo` and `<- foo` was supposed to replace this. So maybe try removing the `- list.map_of_fn` and see if it still works.

Reply:
I can confirm it doesn't work without the `- list.map_of_fn`. I hadn't `squeeze_simp`ed it since I get `arguments of the form '←...' are not supported`. I `squeeze_simp`ed it using the `.symm` version and, oddly enough, the lemma `list.map_of_fn` is not in the final list!

Reply:
@Vierkantor you implemented the reverse simp lemmas, right? Any thoughts here? Not relevant to this PR anymore, but I'm surprised it's ever necessary to write both <- and - for the same simp lemma. There's no situation where you could want <- without -, right?

Reply:
There is code in `simp` to remove `foo` when `<- foo` is present ([relevant commit](https://github.com/leanprover-community/lean/pull/100/commits/cc65428f6d85e39310c06773256a1af14864013e)), so it looks like a core bug. I can't see anything obviously wrong here. Perhaps the two `list.map_of_fn` aren't considered equal for some reason?

Reply:
https://github.com/leanprover-community/lean/issues/193